src/Xrefcheck/Data/Redirect.hs

```diff
@@ -0,0 +1,162 @@
+{- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Xrefcheck.Data.Redirect
+  ( RedirectChain
+  , pushRequest
+  , hasRequest
+  , totalFollowed
+
+  , RedirectRule (..)
+  , RedirectRuleOn (..)
+  , RedirectRuleOutcome (..)
+  , redirectRule
+
+  , isPermanentRedirectCode
+  , isRedirectCode
+  , isTemporaryRedirectCode
+  ) where
+
+import Universum
+
+import Data.Aeson (genericParseJSON)
+import Data.Yaml (FromJSON (..), withText)
+import Fmt (Buildable (..))
+import Text.Regex.TDFA.Text (Regex)
+
+import Xrefcheck.Scan ()
+import Xrefcheck.Util
+
+-- | A custom redirect rule.
+data RedirectRule = RedirectRule
+  { rrFrom :: Maybe Regex
+    -- ^ Redirect source links that match to apply the rule.
+    --
+    -- 'Nothing' matches any link.
+  , rrTo :: Maybe Regex
+    -- ^ Redirect target links that match to apply the rule.
+    --
+    -- 'Nothing' matches any link.
+  , rrOn :: Maybe RedirectRuleOn
+    -- ^ HTTP code selector to apply the rule.
+    --
+    -- 'Nothing' matches any code.
+  , rrOutcome :: RedirectRuleOutcome
+    -- ^ What to do when an HTTP response matches the rule.
+  } deriving stock (Generic)
+
+-- | Rule selector depending on the response HTTP code.
+data RedirectRuleOn
+    = RROCode Int
+      -- ^ An exact HTTP code
+    | RROPermanent
+      -- ^ Any HTTP code considered as permanent according to 'isPermanentRedirectCode'
+    | RROTemporary
+      -- ^ Any HTTP code considered as permanent according to 'isTemporaryRedirectCode'
+    deriving stock (Show, Eq)
+
+-- | What to do when receiving a redirect HTTP response.
+data RedirectRuleOutcome
+    = RROValid
+      -- ^ Consider it as valid
+    | RROInvalid
+      -- ^ Consider it as invalid
+    | RROFollow
+      -- ^ Try again by following the redirect
+    deriving stock (Show, Eq)
+
+-- | Links in a redirection chain.
+--
+-- They are stored in reverse order, so it is quick to
+-- accumulate new ones while following a redirection chain.
+newtype RedirectChain = RedirectChain
+  { unRedirectChain :: [Text]
+  } deriving newtype (Show, Eq)
+
+instance FromList RedirectChain where
+  type ListElement RedirectChain = Text
+  fromList = RedirectChain . reverse
+
+pushRequest :: RedirectChain -> Text -> RedirectChain
+pushRequest (RedirectChain chain) = RedirectChain . (: chain)
```

Comment:
In all these functions, let's replace `Text` with something more meaningful (either `type` or `newtype`)